runtime/memprof.c

```diff
@@ -451,77 +616,190 @@ void caml_memprof_track_young(tag_t tag, uintnat wosize, int from_caml)
-     selecting blocks. [intern.c] will call [check_urgent_gc] which
-     will call [caml_memprof_handle_postponed] in turn. */
   p = block;
-  while(1) {
+  while (1) {
     uintnat next_sample = mt_generate_geom();
     header_t *next_sample_p, *next_p;
-    if(next_sample > blockend - p)
+    if (next_sample > blockend - p)
       break;
     /* [next_sample_p] is the block *following* the next sampled
        block! */
     next_sample_p = p + next_sample;
 
-    while(1) {
+    while (1) {
       next_p = p + Whsize_hp(p);
-      if(next_p >= next_sample_p) break;
+      if (next_p >= next_sample_p) break;
       p = next_p;
     }
 
-    register_postponed_callback(
-      Val_hp(p), mt_generate_binom(next_p - next_sample_p) + 1,
-      Unmarshalled, &callstack);
-
+    if (callstack == 0) callstack = capture_callstack_postponed();
+    if (callstack == 0) break;  /* OOM */
+    new_tracked(mt_generate_binom(next_p - next_sample_p) + 1,
+                Hd_hp(p), 1, is_young, Val_hp(p), callstack);
     p = next_p;
   }
+  caml_memprof_check_action_pending();
+}
+
+/**** Interface with the OCaml code. ****/
+
+#define MT_STATE_SIZE 624
+static void caml_memprof_init(void) {
+  uintnat i;
+
+  init = 1;
+
+  mt_index = MT_STATE_SIZE;
+  mt_state[0] = 42;
+  for (i = 1; i < MT_STATE_SIZE; i++)
+    mt_state[i] = 0x6c078965 * (mt_state[i-1] ^ (mt_state[i-1] >> 30)) + i;
+
+  callback_alloc_minor = Val_unit;
+  callback_alloc_major = Val_unit;
+  callback_promote = Val_unit;
+  callback_dealloc_minor = Val_unit;
+  callback_dealloc_major = Val_unit;
+
+  caml_register_generational_global_root(&callback_alloc_minor);
+  caml_register_generational_global_root(&callback_alloc_major);
+  caml_register_generational_global_root(&callback_promote);
+  caml_register_generational_global_root(&callback_dealloc_minor);
+  caml_register_generational_global_root(&callback_dealloc_major);
+}
+
+void caml_memprof_shutdown(void) {
+  init = 0;
+  lambda = 0.;
+  caml_memprof_suspended = 0;
+  trackst.len = 0;
+  trackst.callback = trackst.young = trackst.delete = 0;
+  caml_stat_free(trackst.entries);
+  trackst.entries = NULL;
+  trackst.alloc_len = 0;
+}
+
+CAMLprim value caml_memprof_set(value lv, value szv,
+                                value cb_alloc_minor, value cb_alloc_major,
+                                value cb_promote,
+                                value cb_dealloc_minor, value cb_dealloc_major)
+{
+  CAMLparam5(lv, szv, cb_alloc_minor, cb_alloc_major, cb_promote);
+  CAMLxparam2(cb_dealloc_minor, cb_dealloc_major);
+  double l = Double_val(lv);
+  intnat sz = Long_val(szv);
+  uintnat i;
+
+  if (sz < 0 || !(l >= 0.) || l > 1.) /* Checks that [l] is not NAN. */
+    caml_invalid_argument("caml_memprof_set");
+
+  /* This call to [caml_memprof_set] will discard all the previously
+     tracked blocks. We try one last time to call the postponed
+     callbacks. */
+  caml_raise_if_exception(caml_memprof_handle_postponed_exn());
```

Comment:
With this code, if a leftover callback raises, then the actual setting changes will not happen. I find it a bit surprising, and I don't think that readers of the `Memprof.{start,stop}` documentation will understand this.

As mentioned previously, I think it would be more natural to handle this as in a a "memprof shutdown" scenario, calling all callbacks and not-raising exceptions. At least, if we want to raise one of the exceptions, we could consider doing it at the end of the function, after changing the settings?

Reply:
To me, ignoring exceptions is not an option: they may be related to actual bugs which must be returned to the user.

Now, we can either decide to re-raise them to the rest of the program, or to crash the program, or to show a warning. I don't really care, but we should so this in a uniform way in shutdown scenarii, in `Memprof.{start,stop}` and in the main program. Note that @gadmm imagined some use cases of callbacks raising exceptions.

About `Memprof.{start,stop}` not setting when a callback raises an exception: I think actually this is the right behavior. Memprof callbacks are asynchronous, which means that they can be called anywhere. Why not say that the exception-raising calback is called at the very moment `Memprof.{start,stop}` is called? Since `Memprof.{start,stop}` has not executed when the exception is raised, it is natural that it does not set the memprof parameters. On the other hand, I don't think there is any serializable execution where the memprof parameters are set and the callback is called, raising an exception.

Reply:
Let me prefix my comment below with: I think it is perfectly reasonable to assume that most callbacks will not raise synchronous exceptions, and the distinction we are discussing does not matter in this scenario, so it would be perfectly reasonable to merge the current PR and possibly iterate more on this later. (I'll go ahead and add an approval stamp.)

There are two separate parts where I recommended the same approach:
- program shutdown time
- `caml_memprof_set` time (change of memprof parameters/callbacks)

(The runtime does ignores exceptions raised by `at_exit` callbacks at shutdown time (see `do_at_exit` usage points; on the other hand, I think the currently implementation stops at the first exception-raising callback and does not force the others).)

The use-case I have in mind for calling all memprof callbacks at shutdown time is a scenario where callbacks manipulate (possibly disk-backed) structures collecting memory-usage statistics. If we flush all callbacks, we give better guarantee that those statistics will present all allocations upto the point of error. That does lead to ignoring some failures, but we do not know how to deal with them at program-termination point anyway. (Would you feel better about at least returning the first failure?)

At `caml_memprof_set` time, it seems natural to me to make a best-effort attempt at calling all callbacks, but your argument regarding propagating exceptions also makes sense.

In a perfect world, I think that what we want is support for collecting several asynchronous exceptions and raising them simultaneously. A poor-man approach to this is to have a `exception Memprof_callbacks of exception list` constructor, as a generalisation of @gadmm's `Finally_raised` wrapper in Fun.protect. If we had separate language constructs for synchronous and asynchronous exceptions, it could even make sense to have a language primitive for "simultaenous" asynchronous raising/propagation, and demand that asynchronous catch-handlers process the simultaneous exceptions in some order. Then the right behaviour for flushing callbacks would always be to process them all, collecting exceptions (asynchronous by definition), and raising them simultaneously.

Reply:
Regarding the usage scenarios of collecting statistics that should be complete: if a user uses `at_exit` to call a function to persist those statistics to the disk, then we want memprof callbacks to be flushed *before* this function runs. This is not supported by the approach of flushing callbacks in `at_exit`. (If we changed the runtime to do this correctly, we would get the guarantee that at_exit callbacks are never monitored, which sounds like a slightly nicer specification.)

Reply:
* `at_exit` is very fragile, so I find it better do it on the C side of caml_shutdown.
* It does not shock me if asynchronous callbacks are dropped given that we can decide they happen _after_ the end of the program (or after Memprof was stopped). I agree with Jacques-Henri that _if_ they ran and raised an exception, then the exception should not be ignored.
* It is indeed perfectly reasonable to assume that most callbacks will not encounter an error, but not happening very often is something we always expect about errors. This is precisely why it is important that the current treatment is already good enough.
* With masking I will probably need to drop those callbacks anyway in `memprof_set` if it is called while the mask is on.
* I am convinced by Jacques-Henri's explanation about `memprof_set` not acting if an exception is raised. Keep in mind that being a stateful interface, correct handling of memprof state will require to treat it as a resource, and thus in a context where callbacks can raise errors or other asynchronous exceptions, `Memprof.stop` will need to be called while the mask is set. So, combined with the previous bullet point, in practice, those callbacks _will_ be dropped, unless the programmer has a model where memprof callbacks cannot fail and they call `Memprof.stop` naked.

One solution is therefore to simply drop callbacks in `memprof_set` right away, but perform the best-effort calling of memprof callbacks suggested by Gabriel just before calling `Memprof.stop` in `caml_shutdown`, stopping at the first exception.

As for the idea collecting multiple asynchronous exceptions... I am sceptical of too complex solutions. In .NET they have something called AggregateException that represents several exceptions at once, but I wonder how they program with it. I imagine it useful if a human is involved in the loop. We already know that asynchronous exceptions cannot have a payload since we cannot reliably distinguish them (cf. `Fun.Finally_raised`). In fact, this is something that Milner, Tofte and Harper already realised. I recommend anybody interested in asynchronous exceptions to read about Interrupts in Standard ML, up to its treatment in Isabelle/ML (cf. http://isabelle.in.tum.de/dist/library/Doc/Implementation/ML.html, Ctrl+F "Interrupts"). OCaml misses at least an `is_interrupt` predicate, a rule that asynchronous exceptions are interrupts, a rule that one must not discriminate interrupts, and a rule that an interrupt must always be re-raised promptly except at isolation boundaries. If furthermore there was a dedicated Interrupt exception then it would make sense if it had a payload that collects several errors for reporting, on the condition that the programmer should not match on those errors programmatically, and knowing that there cannot be a guarantee of exhaustiveness given that it can be replaced by any interrupt arising later.

Reply:
Re. dropping callbacks, my general view is that callbacks that correspond to allocations that happen before a particular event (either `caml_memprof_set` or the decision to terminate the program) should ideally be called rather than dropped, so that the profiler-data consumer has a complete view of the memory usage that happened before this event.

Dropping callbacks that come after an exception-in-a-callback seems a bit strange to me; as far as I know the callbacks are not currently guaranteed to be called in allocation order (it would be nice if we could give this guarantee), so this could result in "forgetting" some allocation errors whose sequential order is independent to the one whose callback failed. I would find it more natural to call all callbacks in all cases, and raise one of the exceptions (and drop the others; we cannot have everything).

In particular, if some callbacks are raising synchronous exceptions due to programming errors (instead of propagating asynchronous exception as it is natural they would), the behavior proposed above does provide evidence to the user that some programming error(s) happened, it's just that they do not get all the error reports.

Reply:
Let us assume the programmer wants a complete view as @gasche described, and try to think about how they can call `Memprof.stop` reliably _in the presence of asynchronous exceptions_. Long story short, this is a case for [Fun.with_resource](https://github.com/ocaml/ocaml/pull/8962). `Memprof.stop` will run inside `release`, which cannot propagate any exception, so either one drops callbacks or one drops possible exceptions (and then calls again `Memprof.stop`). In either situation, the programmer seeing that might prefer to flush the callbacks by hand at the end of the success paths they care about. This way, they have the best of both worlds: they both have a complete view, and retain their ability to reason about their program.

This is very similar to another issue: “how can you ignore exceptions coming out of `close_in/out` inside `release`?” Distinguish _releasing_ from _committing_. Resource release must be prompt and cannot fail: it only cares about cleaning up world state. Committing can take time and fail (e.g. calling flush/fsync in a place where you are ready to handle an exception). Committing can be done explicitly at the end of success paths, which are few and known to the programmer, and where once is able to handle failure as part of the program logic. If the program is interrupted in the middle of the work, I might prefer for instance best-effort flushing (for a log file) or removing the corrupted file (for a temporary file).

I think this is a common pattern addressing a common issue (and a particular case of the "double raise" issue): it is a fundamental impossibility to guarantee both resource safety, and allow handling of committing errors as part of the program logic in every unexpected failure paths.

So, the best is an API that allows a clear distinction between releasing and committing. I will take care of this when I will implement masking, where I will have no other choice than dropping callbacks in `memprof_set`. It is ok for me to propose an imperfect but convenient behaviour as the default (e.g. when masking is not on) as long as the expert programmer can still get what they need (the file comparison is useful here again: OCaml has made the choice of best-effort flushing and ignoring errors at program exit and in `close_out_noerr` for the sake of convenience/beginner friendliness). What I retain from @gasche's point is that there is also an interest in having some function `Memprof.flush` that guarantees there are no more callbacks if it returns without raising.

> @gasche: as far as I know the callbacks are not currently guaranteed to be called in allocation order (it would be nice if we could give this guarantee),

I agree this would seem a natural guarantee to have @jhjourdan.

> @gasche: if some callbacks are raising synchronous exceptions

What do you mean with synchronous exception @gasche? To me, any exception raised by an asynchronous callback is asynchronous by definition.

Reply:
Exceptions raised out of callbacks are asynchronous. But from the point of view of the callbacks  themselves, they may raise synchronous or asynchronous exceptions -- which are then seen as asynchronous from the rest of the world. (But then maybe you assimilate all programming-errors exceptions to asynchronous exceptions, in the sense that they do not come from the intended normal control flow of the program; then the distinction is moot.)

Reply:
We may, medium-term, have a form of "default memprof implementation" that one may enable for arbitrary OCaml programs without modifying their codebase. (Similar to 24e42d4d227b884894d9e7f4639c591bfe9d5496, which  ). I agree that giving clear control to expert Memprof users is useful, but it would be nice to also have good guarantees in such a "default" mode.

Reply:
The default mode could make sure to handle exceptions inside the callbacks for instance, so that it does not interfere with the program.

Reply:
The constraints on handling programming errors are already very close to the ones for asynchronous exceptions, and the current practice of allowing raising Failure or Assert_failure from finalisers means they are already asynchronous to some programmers. For the semantics of `Fun.Finally_raised` they are already indistinguishable. So yes, I (secretly) assimilate programming errors to asynchronous exceptions, to be handled only at isolation boundaries.